tests/test_packages/test_skills/test_tac_control/test_behaviours.py

```diff
@@ -0,0 +1,700 @@
+        self.parameters._end_time = mocked_end_time
+
+        datetime_mock = Mock(wraps=datetime.datetime)
+        datetime_mock.now.return_value = mocked_now_time
+
+        self.parameters._min_nb_agents = 2
+        self.game._registration.register_agent(COUNTERPARTY_ADDRESS, "agent_name_1")
+        mocked_description = Description({"foo1": 1, "bar1": 2})
+
+        dialogue = self.prepare_skill_dialogue(
+            self.tac_dialogues,
+            (
+                DialogueMessage(
+                    TacMessage.Performative.REGISTER,
+                    {"agent_name": "some_agent_name"},
+                    True,
+                ),
+            ),
+        )
+        dialogue._incoming_messages = []
+
+        # operation
+        with patch("datetime.datetime", new=datetime_mock):
+            with patch.object(
+                self.game,
+                "get_unregister_tac_description",
+                return_value=mocked_description,
+            ):
+                with patch.object(self.tac_behaviour.context.logger, "log"):
+                    with pytest.raises(
+                        ValueError, match="Error when retrieving last message."
+                    ):
+                        self.tac_behaviour.act()
+
+    def _assert_tac_message_and_logging_output(
+        self, tac_message: TacMessage, participant_address: Address, mocked_logger,
+    ):
+        has_attributes, error_str = self.message_has_attributes(
+            actual_message=tac_message,
+            message_type=TacMessage,
+            performative=TacMessage.Performative.GAME_DATA,
+            to=participant_address,
+            sender=self.skill.skill_context.agent_address,
+        )
+        assert has_attributes, error_str
+        mocked_logger.assert_any_call(
+            logging.DEBUG,
+            f"sending game data to '{participant_address}': {str(tac_message)}",
+        )
+
+    def test_act_iii(self):
+        """Test the act method of the tac behaviour where phase is game_registration and start_time < now < end_time and nb_agent < min_nb_agents"""
+        # setup
+        self.game._phase = Phase.GAME_REGISTRATION
+
+        mocked_start_time = datetime.datetime.strptime(
+            "01 01 2020  00:01", "%d %m %Y %H:%M"
+        )
+        mocked_end_time = datetime.datetime.strptime(
+            "01 01 2020  00:03", "%d %m %Y %H:%M"
+        )
+        mocked_now_time = datetime.datetime.strptime(
+            "01 01 2020  00:02", "%d %m %Y %H:%M"
+        )
+
+        self.parameters._start_time = mocked_start_time
+        self.parameters._end_time = mocked_end_time
+
+        datetime_mock = Mock(wraps=datetime.datetime)
+        datetime_mock.now.return_value = mocked_now_time
+
+        self.parameters._min_nb_agents = 2
+        agent_1_address = "agent_address_1"
+        agent_1_name = "agent_name_1"
+        agent_2_address = "agent_address_2"
+        agent_2_name = "agent_name_2"
```

Comment:
some repeat variables across tests could be reused

Reply:
Done